conmon-rs/server/src/attach.rs

```diff
@@ -192,74 +203,96 @@ impl Attach {
-                    ),
-                },
-                _ => {}
             }
         }
     }
 
     async fn write_loop(
         mut write_half: OwnedWriteHalf,
         mut rx: Receiver<(Pipe, Vec<u8>)>,
+        token: CancellationToken,
     ) -> Result<()> {
         loop {
-            let (pipe, buf) = rx.recv().await?;
-
-            let mut packets = buf
-                .chunks(Self::PACKET_BUF_SIZE - 1)
-                .map(|x| {
-                    let mut y = x.to_vec();
-                    let p = match pipe {
-                        Pipe::StdOut => 2,
-                        Pipe::StdErr => 3,
-                    };
-                    y.insert(0, p);
-                    y.resize(Self::PACKET_BUF_SIZE, 0);
-                    y
-                })
-                .collect::<Vec<_>>();
-            packets.push(Self::DONE_PACKET.to_vec());
+            select! {
+                res = rx.recv() => {
+                    let (pipe, buf) = res?;
+                    let packets = buf
+                        .chunks(Self::PACKET_BUF_SIZE - 1)
+                        .map(|x| {
+                            let mut y = x.to_vec();
+                            let p = match pipe {
+                                Pipe::StdOut => 2,
+                                Pipe::StdErr => 3,
+                            };
+                            y.insert(0, p);
+                            y.resize(Self::PACKET_BUF_SIZE, 0);
+                            y
+                        })
+                        .collect::<Vec<_>>();
 
-            let len = packets.len() - 1;
-            for (idx, packet) in packets.iter().enumerate() {
-                match write_half.write(packet).await {
-                    Ok(_) => {
-                        debug!("Wrote {} packet {}/{} to client", pipe, idx, len)
+                    let len = packets.len() - 1;
+                    for (idx, packet) in packets.iter().enumerate() {
+                        match write_half.write(packet).await {
+                            Ok(_) => {
+                                debug!("Wrote {} packet {}/{} to client", pipe, idx, len)
+                            }
+                            Err(ref e) if e.kind() == ErrorKind::WouldBlock => continue,
+                            Err(ref e) if e.kind() == ErrorKind::BrokenPipe => break,
+                            Err(e) => bail!("unable to write packet {}/{}: {:#}", idx, len, e),
+                        }
                     }
-                    Err(ref e) if e.kind() == ErrorKind::WouldBlock => continue,
-                    Err(ref e) if e.kind() == ErrorKind::BrokenPipe => break,
-                    Err(e) => bail!("unable to write packet {}/{}: {:#}", idx, len, e),
+                }
+                _ = token.cancelled() => {
+                    debug!("Exiting because token cancelled");
+                    match write_half.write(Self::DONE_PACKET).await {
```

Comment:
should this be a CANCELLED_PACKET ?

Reply:
I don't think we need to make the distinction. either way, the client will disconnect with no error (unless the exit code was non-zero, which is communicated with a different mechanism